File: cpp/PrintReversedList.cpp
/*
* Title: Print reverse list
*
* Execution: g++ PrintReverseList.cpp -o PrintReverseList
*
* For more details, check out http://www.byte-by-byte.com/printreverselist/
*/

#include <iostream>
#include <vector>

using namespace std;

struct Node {
    int value;
    Node *next;
};

void printReverseList(Node *n) {
    if (n == NULL) {
        return;
    }
    printReverseList(n->next);
    cout << n->value << endl;
}

void test_1() {
    
    cout << "Original list: 1->2->3" << endl;
    Node *n1 = new Node();
    Node *n2 = new Node();
    Node *n3 = new Node();

    n1->value = 1;
    n1->next = n2;

    n2->value = 2;
    n2->next = n3;

    n3->value = 3;
    n3->next = NULL;

    printReverseList(n1);
}

void test_2() {

    cout << "Original list: 3->2->1" << endl;
    Node *n1 = new Node();
    Node *n2 = new Node();
    Node *n3 = new Node();

    n1->value = 3;
    n1->next = n2;

    n2->value = 2;
    n2->next = n3;

    n3->value = 1;
    n3->next = NULL;

    printReverseList(n1);
}

void test_3() {

    cout << "Original list: 10->9" << endl;
    Node *n1 = new Node();
    Node *n2 = new Node();

    n1->value = 10;
    n1->next = n2;

    n2->value = 9;
    n2->next = NULL;

    printReverseList(n1);
}

int main() {

    test_1();
    test_2();
    test_3();

    return 0;
}

File: java/PrintReversedList.java
/*
 *   Title: Print reversed list
 *
 *   Given a linked list, write a function that prints the nodes of the 
 *   list in reverse order.
 * 
 *   Execution: javac PrintReversedList.java && java PrintReversedList
 * 
 *   For more details, check out http://www.byte-by-byte.com/printreversedlist/
 */

public class PrintReversedList {

    private static class Node {
        private int value;
        private Node next;
        
        private Node(int value) {
            this.value = value;
        }
    }

    public static void printReversedList(Node n) {
        if (n == null) {
            return;
        }
        printReversedList(n.next);
        System.out.println(n.value);
    }

    public static void test_1() {
        System.out.println("Original: 1->2->3->4");
        Node n = new Node(1);
        n.next = new Node(2);
        n.next.next = new Node(3);
        n.next.next.next = new Node(4);

        printReversedList(n);
    }

    public static void test_2() {
        System.out.println("Original: 4->3->2->1");
        Node n = new Node(4);
        n.next = new Node(3);
        n.next.next = new Node(2);
        n.next.next.next = new Node(1);

        printReversedList(n);
    }
    
    public static void test_3() {
        System.out.println("Original: 10->9->8");
        Node n = new Node(10);
        n.next = new Node(9);
        n.next.next = new Node(8);

        printReversedList(n);
    }
    // Sample test cases
    public static void main(String[] args) {

        System.out.println("Test 1:");
        test_1();
        System.out.println("\n");

        System.out.println("Test 2:");
        test_2();
        System.out.println("\n");

        System.out.println("Test 3:");
        test_3();
        System.out.println("\n");

        System.out.println("Passed all test cases");
    }
}
